### PR TITLE
Implement Quad-Buffered stereoscopy support

### PR DESCRIPTION
**Pull request by Armada651**
Quad-Buffering is a native stereoscopy API that doesn't use hacks and allows use to directly output our layered EFB to the backbuffer.

For D3D11 this requires an upgrade to DXGI 1.2, thus this makes the D3D backend incompatible with Vista. Also, because we don't ever recreate the swapchain for D3D we always keep the stereo flag enabled on the swapchain. This allows use to toggle Quad-Buffered stereoscopy during gameplay and there doesn't seem to be a downside to keeping that flag enabled even if no 3D display modes are available.

In the Video Config dialog I called Quad-Buffering "HDMI 3D", because this functionality is often associated with the HDMI 1.4a standard that adds support for 3D display modes.

This is a new attempt of PR #1792.

### CODE REVIEW

**Comment by riking** [comment]
RIP Dolphin on Vista 😢

**Comment by MayImilae** [comment]
@riking OpenGL? Besides, it hasn't been officially supported for some time! :3

**Comment by fantesykikachu** [comment]
Could we call it AMD HD3D like you said in the original PR so as to be less confusing for HDMI users, and i feel like we don't need a description because 3D vision doesn't have one. Also I tested it with my 3D monitor and most games ran fine, there are a few games that don't like 3D and break no matter which 3D mode you set them in, however the one issue I did notice is that if run a combination of "HDMI 3D", OpenGL, and any anti-aliasing; no video is displayed (audio and controls still work), and these settings seem to do this on non-3D monitors too. I also tested DX11 with normal settings (no 3D) and there seemed to be a slight improvement and no real regressions, however i didn't spend very long testing that part.

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 1]
```diff
@@ -324,45 +293,34 @@ HRESULT Create(HWND wnd)
     UpdateActiveConfig();
   }
 
-  DXGI_SWAP_CHAIN_DESC swap_chain_desc = {};
-  swap_chain_desc.BufferCount = 1;
+  DXGI_SWAP_CHAIN_DESC1 swap_chain_desc = {};
+  swap_chain_desc.BufferCount = NUM_SWAPCHAIN_BUFFERS;
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: Is this an implicit change of the behavior from vsync to tripple buffering?

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 2]
```diff
@@ -324,45 +293,34 @@ HRESULT Create(HWND wnd)
...
-  if (swap_chain_desc.Windowed)
-  {
-    // forcing buffer resolution to xres and yres..
-    // this is not a problem as long as we're in windowed mode
-    swap_chain_desc.BufferDesc.Width = xres;
-    swap_chain_desc.BufferDesc.Height = yres;
-  }
+  // By always creating a stereo swapchain we can toggle Quad-Buffered stereoscopy
+  // while the game is running.
+  swap_chain_desc.Stereo = TRUE;
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: Will this increase the GPU load while stereo is not in use?

**Comment by degasus** [comment]
As there is no reflection if a stereo swapchain was created successfully, what will happen if the host doesn't support it?

File: Source/Core/VideoBackends/D3D/D3DBase.cpp (hunk 1 above)
Comment (Armada651) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: This is an implicit change for a flipping swap-chain which is required for Quad-Buffered stereoscopy. Setting the BufferCount to 1 actually means we were doing front-buffered rendering in exclusive fullscreen:

> When you call IDXGIFactory::CreateSwapChain to create a full-screen swap chain, you typically include the front buffer in this value.

> In full-screen mode, there is a dedicated front buffer; in windowed mode, the desktop is the front buffer.

File: Source/Core/VideoBackends/D3D/D3DBase.cpp (hunk 2 above)
Comment (Armada651) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: That's not documented anywhere, we will need to do some performance testing if we want to know that. But logically speaking it should only result in the allocation of a larger backbuffer.

**Comment by Armada651** [comment]
@degasus If the host doesn't support it the stereoscopy effect will just not be displayed, though we still render it internally.

We could use this function to see if the user is actually using a display mode that supports stereoscopy right now: https://msdn.microsoft.com/en-us/library/windows/desktop/hh404561(v=vs.85).aspx

In that case we could display a warning message on the OSD.

**Comment by degasus** [comment]
I think this would be a good step, but LGTM for now ;)

**Comment by Armada651** [comment]
@fantesykikachu The problem with calling it "AMD HD3D" is that it is no longer limited to just AMD, apparently Nvidia users support it too.

Calling it "HDMI 3D" is not ideal, but if people google that term they definitely get more helpful information than if I call it "Quad-Buffer".

I'll look into the OpenGL + MSAA + Stereo issue, though that may have been an issue before this PR.

**Comment by Armada651** [comment]
@fantesykikachu PR #5712 fixes the problem with OpenGL Stereoscopy that you've reported.

**Comment by mimimi085181** [comment]
5.0-4657 breaks D3D, 5.0-4647 works fine.

"Failed to initialize Direct3D. Make sure your video card supports at least D3D 10.0"

**Pull request merged**